torch/_ops.py

```diff
@@ -127,6 +127,110 @@ def inner(fn):
 
         return inner
 
+    # Registers an implementation to all **3** variants of functionalization that we have:
+    # - DispatchKey.Functionalize
+    # - functorch.TransformType.FuncTorchFunctionalize
+    # - FunctionalTensorMode
+    def py_functionalize_impl(self):
+        from torch._C import _functionalization_reapply_views_tls as _reapply_views
+        from torch._functorch.eager_transforms import (
+            _unwrap_all_tensors_from_functional,
+            _wrap_all_tensors_to_functional,
+        )
+        from torch._subclasses.functional_tensor import (
+            dispatch_functionalize,
+            FunctionalTensor,
+            FunctionalTensorMode,
+            unset_functional_temporarily,
+        )
+
+        class _DictToClass:
+            def __init__(self, d):
+                for k, v in d.items():
+                    setattr(self, k, v)
+
+        functionalize_dk_methods = {
+            "unwrap": lambda args: _unwrap_all_tensors_from_functional(
+                args, reapply_views=_reapply_views()
+            ),
+            "wrap": lambda args: _wrap_all_tensors_to_functional(args, level=0),
+            "wrap_function": lambda f: torch.func.functionalize(
+                f, remove="mutations_and_views" if _reapply_views() else "mutations"
+            ),
+            "redispatch": lambda: torch._C._ExcludeDispatchKeyGuard(
+                torch._C.DispatchKeySet(torch._C.DispatchKey.Functionalize)
+            ),
+        }
+
```

Comment:
For code readability: could we author these dicts as classes instead of dicts?  That is, there is some BaseFunctionalizeAPI class with four methods: wrap, unwrap, wrap_function, redispatch, then FunctorchFunctionalizeAPI, PythonFunctionalizeAPI, and CppFunctionalizeAPI can subclass from it

Also, at the risk of complicating your PR stack: something that might help is moving HigherOrderOp out of torch/_ops.py because every global member of torch/_ops.py becomes torch.ops.blah (and therefore public API)

Reply:
> For code readability: could we author these dicts as classes instead of dicts? 

Sure - I originally do the dict-to-class approach just to minimize boilerplate, but happy to make them actual classes for readability.

> Also, at the risk of complicating your PR stack: something that might help is moving HigherOrderOp out of torch/_ops.py because every global member of torch/_ops.py becomes torch.ops.blah (and therefore public API)

Hmm. This is totally fair. Just to minimize the changes in the PR, are you ok with me authoring the new classes in `torch/_subclasses/functional_tensor.py` and importing them into here? I agree that moving `HigherOrderOperator` out of `torch/_ops.py` feels like the right thing though.

Reply:
> Hmm. This is totally fair. Just to minimize the changes in the PR, are you ok with me authoring the new classes in torch/_subclasses/functional_tensor.py and importing them into here? I agree that moving HigherOrderOperator out of torch/_ops.py feels like the right thing though.

Yes, as long as we confirm that importing them into torch._ops doesn't make them public. The torch._ops -> torch.ops thing makes me nervous

Reply:
agreed - i'll confirm that this is the case